### PR TITLE
Num cpus

### PR DESCRIPTION
**Pull request by danieleades**
depends on #3 

this pull request adds an opt-in feature to set the number of threads automatically.

It also restructures the crate so that it bundles it re-exports its dependencies (`smol`, optionally `num_cpus`), rather than relying on the user to add them manually.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1 @@
 /target
-Cargo.lock
```
Comment (wusyong) [inline .gitignore:2]: Could you bring this back? I would prefer default setting on cargo.

File: smol-potat-macro/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+/target
+Cargo.lock
```
Comment (wusyong) [inline smol-potat-macro/.gitignore]: This file isn't needed I think.

**Review by wusyong — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (danieleades) [reply .gitignore:2]: I can take a look at this when I get home. I think in general the usual approach is to let the application author commit the cargo.lock, and to *not* specify concrete versions in *libraries* (such as this one). 
https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html

File: smol-potat-macro/.gitignore (hunk 2 above)
Comment (danieleades) [reply smol-potat-macro/.gitignore]: done

File: .gitignore (hunk 1 above)
Comment (danieleades) [reply .gitignore:2]: here's a direct link to this exact issue in the FAQ - https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries

tl;dr: the cargo.lock is ignored unless the crate is at the end of the dependency chain (which this crate *never* is).

Comment (wusyong) [reply .gitignore:2]: Okay, thanks! But I think it might be better for me to generate this lock file with verified commit myself. I will merge this and push it later.

**Review by wusyong — APPROVED** [review]

File: .gitignore (hunk 1 above)
Comment (Restioson) [reply .gitignore:2]: There is no reason to keep committing it if is irrelevant then, surely? It's just extra stuff to track in git and inflates the lines-changed count in commits

Comment (danieleades) [reply .gitignore:2]: Hi @Restioson, this conversation is already resolved and this pull request is already merged. If you take a look at the commits you'll see that this was file was *not* tracked as part of this pull request.

Comment (Restioson) [reply .gitignore:2]: OK :)
